homeassistant/components/mqtt/config_flow.py

```diff
@@ -123,6 +144,163 @@ async def async_step_hassio_confirm(self, user_input=None):
+                self.broker_config.update(user_input)
+                return await self.async_step_options()
+
+            errors["base"] = "cannot_connect"
+
+        fields = OrderedDict()
+        fields[vol.Required(CONF_BROKER, default=current_config[CONF_BROKER])] = str
+        fields[vol.Required(CONF_PORT, default=current_config[CONF_PORT])] = vol.Coerce(
+            int
+        )
+        fields[
+            vol.Optional(
+                CONF_USERNAME,
+                description={"suggested_value": current_config.get(CONF_USERNAME)},
+            )
+        ] = str
+        fields[
+            vol.Optional(
+                CONF_PASSWORD,
+                description={"suggested_value": current_config.get(CONF_PASSWORD)},
+            )
+        ] = str
+
+        return self.async_show_form(
+            step_id="broker", data_schema=vol.Schema(fields), errors=errors,
+        )
+
+    async def async_step_options(self, user_input=None):
+        """Manage the MQTT options."""
+        errors = {}
+        current_config = self.config_entry.data
+        options_config = {}
+        if user_input is not None:
+            bad_birth = False
+            bad_will = False
+
+            if "birth_topic" in user_input:
+                birth_message = {
+                    ATTR_TOPIC: user_input["birth_topic"],
+                    ATTR_PAYLOAD: user_input.get("birth_payload", ""),
+                    ATTR_QOS: user_input["birth_qos"],
+                    ATTR_RETAIN: user_input["birth_retain"],
+                }
+                try:
+                    birth_message = MQTT_WILL_BIRTH_SCHEMA(birth_message)
+                    options_config[CONF_BIRTH_MESSAGE] = birth_message
+                except vol.Invalid:
+                    errors["base"] = "bad_birth"
+                    bad_birth = True
+
+            if "will_topic" in user_input:
+                will_message = {
+                    ATTR_TOPIC: user_input["will_topic"],
+                    ATTR_PAYLOAD: user_input.get("will_payload", ""),
+                    ATTR_QOS: user_input["will_qos"],
+                    ATTR_RETAIN: user_input["will_retain"],
+                }
+                try:
+                    will_message = MQTT_WILL_BIRTH_SCHEMA(will_message)
+                    options_config[CONF_WILL_MESSAGE] = will_message
+                except vol.Invalid:
+                    errors["base"] = "bad_will"
+                    bad_will = True
+
+            options_config[CONF_DISCOVERY] = user_input[CONF_DISCOVERY]
+
+            if not bad_birth and not bad_will:
+                updated_config = {}
+                updated_config.update(self.broker_config)
+                updated_config.update(options_config)
+                self.hass.config_entries.async_update_entry(
+                    self.config_entry, data=updated_config
+                )
+                return self.async_create_entry(title="", data=None)
```

Comment:
I think that this is the wrong approach.

I think that we should create an MQTT panel in the frontend and put this logic inside a WebSocket API. We can then also move the MQTT dev tools over.

Reply:
I like the idea of moving the MQTT Dev tools to the integration card, but I'm not sure why it's better to have MQTT configuration implement its own flow for setting options.
I started a bit in that direction, but I felt I was just reinventing the wheel with error handling, etc.

Reply:
You're right. This solution is okay for now. We can always migrate to the panel later.

Reply:
We could call the `showOptionsFlowDialog` from the MQTT panel?